### PR TITLE
Change clientcore's baseline to Java 17

### PR DESCRIPTION
**Pull request by alzimmermsft**
# Description

Changes the baseline Java version for `clientcore` packages to Java 17 and folds the JDK HttpClient `HttpClient`-based implementation into the core library.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/clientcore/core/src/main/java/io/clientcore/core/util/auth/CompositeChallengeHandler.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package io.clientcore.core.util.auth;
```
Comment (samvaity) [inline sdk/clientcore/core/src/main/java/io/clientcore/core/util/auth/CompositeChallengeHandler.java:3]: nit: line break.
moved to its own file?

Comment (alzimmermsft) [reply sdk/clientcore/core/src/main/java/io/clientcore/core/util/auth/CompositeChallengeHandler.java:3]: I did this as `ChallengeHandler` was exposing a public `ClientLogger` as interfaces cannot have private constants and the logger was only being used by this inner class.

**Review by samvaity — APPROVED** [review]

File: sdk/clientcore/platform-matrix.json [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      "NotFromSource": { "TestFromSource": false }
+    },
+    "RunAggregateReports": {
+      "$": { "RunAggregateReports": false }
+    },
+    "TestOptions": {
+      "$": { "TestOptions": "" }
+    },
+    "TestGoalsGrouping": {
+      "TestsOnly": { "TestGoals": "surefire:test failsafe:integration-test failsafe:verify" }
```
Comment (benbp) [inline sdk/clientcore/platform-matrix.json:24]:
```suggestion
    "TEST_HTTP_CLIENT_IMPLEMENTATION": [ "io.clientcore.core.http.client.DefaultHttpClientProvider", "io.clientcore.http.okhttp3.OkHttpHttpClientProvider" ],
    "Options": {
      "NotFromSource_TestsOnly": {
        "TestFromSource": false,
        "RunAggregateReports": false,
        "TestOptions": "",
        "TestGoals": "surefire:test failsafe:integration-test failsafe:verify"
      }
```

File: sdk/clientcore/platform-matrix.json [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        "$": { "TestFromSource": false }
+      },
+      "RunAggregateReports": {
+        "AggregateReports": { "RunAggregateReports": true }
+      },
+      "TestOptions": {
+        "SkipRebuild": { "TestOptions": "-DskipCompile=true -DskipTestCompile=true -DcreateSourcesJar=false" }
+      },
+      "TestGoalsGrouping": {
+        "Verify": { "TestGoals": "verify" }
```
Comment (benbp) [inline sdk/clientcore/platform-matrix.json:69]: Same grouping consolidation as above.

File: sdk/clientcore/core/src/main/java/io/clientcore/core/http/client/DefaultHttpClient.java [hunk 4]
```diff
@@ -3,333 +3,217 @@
...
             case BUFFER:
             case DESERIALIZE:
                 // Deserialization will occur at a later point in HttpResponseBodyDecoder.
-                if (isTextEventStream(responseHeaders)) {
-                    HttpResponseAccessHelper.setBody(httpResponse, createBodyFromServerSentResult(serverSentResult));
+                if (isTextEventStreamContentType(contentType)) {
+                    body = createBodyFromServerSentResult(serverSentResult);
                 } else {
-                    eagerlyBufferResponseBody(httpResponse, connection.getInputStream());
+                    body = createBodyFromResponse(response);
```
[the remark this answers is not in the document]
Comment (alzimmermsft) [reply sdk/clientcore/core/src/main/java/io/clientcore/core/http/client/DefaultHttpClient.java:189]: It's not, which is what the method being called is doing, where it reads all the bytes of the response into memory eagerly.

File: sdk/clientcore/core/src/main/java/io/clientcore/core/http/client/DefaultHttpClient.java [hunk 5]
```diff
@@ -3,333 +3,217 @@
...
-import io.clientcore.core.http.models.ProxyOptions;
 import io.clientcore.core.http.models.RequestOptions;
 import io.clientcore.core.http.models.Response;
 import io.clientcore.core.http.models.ResponseBodyMode;
 import io.clientcore.core.http.models.ServerSentEventListener;
-import io.clientcore.core.implementation.AccessibleByteArrayOutputStream;
-import io.clientcore.core.implementation.http.HttpResponseAccessHelper;
-import io.clientcore.core.implementation.util.UriBuilder;
-import io.clientcore.core.models.SocketConnection;
-import io.clientcore.core.models.SocketConnectionCache;
```
[the remark this answers is not in the document]
Comment (alzimmermsft) [reply sdk/clientcore/core/src/main/java/io/clientcore/core/http/client/DefaultHttpClient.java:21]: I'll delete them for now

**Review by benbp — APPROVED** [review]

**Pull request merged**